### PR TITLE
merge my v2 to your v2

### PR DESCRIPTION
**Pull request by AdamCoulterOz**
Summary of how this changes your existing PR:
- moves endpoint to provider configuration (have multiple provider blocks if you need to target different app configuration instances, which is standard practice, note this forces a breaking/major version change)
- adds explicit client config to provider configuration (support for `creds`, `cli`, `msi` and `env`). This probably needs more work
- renames a few things to be more standard
- removes context / diagnostic returns (instead of error) as you weren't actually using them; I'm not sure what the official guidance is on this from Terraform though
- adds 5 retries @ 1s starting interval, with exponential backoff

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -5,4 +5,4 @@ azureenv.sh
 .vscode/
 !terraform-provider-akc_v*.exe
 .publish/
-vendor/
\ No newline at end of file
+vendor/
```
Comment (arkiaconsulting) [inline .gitignore:8]: Do we really need to include the `vendor` folder ?

File: akc/provider.go [hunk 2]
```diff
@@ -1,15 +1,41 @@
...
+				Optional:    true,
+				DefaultFunc: schema.EnvDefaultFunc("ARM_CLIENT_ID", nil),
+			},
+			"client_secret": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Sensitive:   true,
+				DefaultFunc: schema.EnvDefaultFunc("ARM_CLIENT_SECRET", nil),
+			},
+			"tenant_id": {
```
Comment (arkiaconsulting) [inline akc/provider.go:28]: I think that the guidelines is now to have the tenant Id sensitive ( at least it is in the AzureCLI pipeline task in Azure DevOps)

**Comment by arkiaman** [comment]
The deal was to use the `go-azure-helpers` package to benefit from its authentication. Because the autorest package requires the credential variables to be `AZURE_......` which is not the terraform azurerm convention (`ARM_.....`).

But the way you implemented it, we might not need it. The pros of the terraform helpers is that it manages the auth method for us.

NB: the `go-azure-helpers` makes use of a context, which is the reason why I introduced it in the terraform methods.

File: akc/provider.go [hunk 3]
```diff
@@ -18,12 +44,34 @@ func Provider() *schema.Provider {
...
+	switch authMethod {
+	case "creds":
+		c, err = client.NewClientCreds(endpoint, clientId, clientSecret, tenantId)
+	case "cli":
+		c, err = client.NewClientCli(endpoint)
+	case "msi":
+		c, err = client.NewClientMsi(endpoint)
+	case "env":
+	default:
+		c, err = client.NewClientEnv(endpoint)
```
Comment (arkiaman) [inline akc/provider.go]: Maybe we could default to `NewClientCli` ?

File: .gitignore (hunk 1 above)
Comment (arkiaman) [inline .gitignore:8]: Do we really need to checkin the `vendor` folder ?

File: akc/provider.go (hunk 2 above)
Comment (arkiaman) [inline akc/provider.go:31]: I think that the guidelines is now to have the tenant Id sensitive ( at least it is in the AzureCLI pipeline task in Azure DevOps)

**Review by arkiaman — CHANGES_REQUESTED** [review]
The next release will be a major version, so I don't have any problem with breaking changes

File: .gitignore (hunk 1 above)
Comment (AdamCoulterOz) [reply .gitignore:8]: I added it to .gitignore, so it won't be checked in my mistake, so I am already in agreement with you 👍🏻

Comment (arkiaman) [reply .gitignore:8]: 🤦

File: akc/provider.go (hunk 2 above)
Comment (AdamCoulterOz) [reply akc/provider.go:31]: done

File: akc/provider.go (hunk 3 above)
Comment (AdamCoulterOz) [reply akc/provider.go]: let me fix it up to follow the same as azurerm:
- `msi` (if `msi=true`)
- `creds` (if set)
- `env` (if set)
- `cli` (default)

Comment (AdamCoulterOz) [reply akc/provider.go]: @arkiaconsulting  done

**Comment by AdamCoulterOz** [comment]
@arkiaconsulting:

> The deal was to use the `go-azure-helpers` package to benefit from its authentication. Because the autorest package requires the credential variables to be `AZURE_......` which is not the terraform azurerm convention (`ARM_.....`).
> 
> But the way you implemented it, we might not need it. The pros of the terraform helpers is that it manages the auth method for us.
> 
> NB: the `go-azure-helpers` makes use of a context, which is the reason why I introduced it in the terraform methods.

Yeah, I can see that, but it looks like `github.com/Azure/go-autorest/autorest/azure/auth` already supports the things we needed?

If you still want to use it, then go ahead, don't let me stop you 👍🏻

**Comment by arkiaman** [comment]
Quoted @AdamCoulterOz

> Yeah, I can see that, but it looks like github.com/Azure/go-autorest/autorest/azure/auth already supports the things we needed?

Let's go as is for now.
I'll checkout your branch, run the acceptance tests and let you informed.

**Comment by AdamCoulterOz** [comment]
Quoted @arkiaconsulting 

> Let's go as is for now.
> I'll checkout your branch, run the acceptance tests and let you informed.

Just to be clear, my branch is off your branch `azure_auth_hashicorp`

File: client/client.go [hunk 4]
```diff
@@ -210,7 +178,11 @@ func (client *AppConfigClient) send(label string, key string, additionalDecorato
 		return nil, UnexpectedError.wrap(err)
 	}
 
-	resp, err := client.Send(req)
+	var retryDecorator autorest.SendDecorator
+	if retry {
+		retryDecorator = autorest.DoRetryForAttempts(5, 1*time.Second)
+	}
+	resp, err := client.Send(req, retryDecorator)
```
Comment (arkiaman) [inline client/client.go:185]: @AdamCoulterOz  adding the decorator make the tests panic (`invalid memory address or nil pointer dereference`) in autorest.DecorateSender, but only on the second request. For example, when running a single test, the test method succeeds but the teardown fails.

Comment (arkiaman) [reply client/client.go:185]: Ok, I've found the issue. We shouldn't pass the decorator if it's not defined

**Comment by arkiaman** [comment]
Thanks @AdamCoulterOz for your contribution.

One thing that annoys me: I think I will revert back to having the endpoint stay on the key-value resources. Otherwise, the provider configuration is dependent on the App Configuration endpoint, that might be provisioned into the terraform configuration. And because of this [this](https://github.com/hashicorp/terraform/issues/2430) a provider configuration cannot reference a resource. Furthermore, that will force us to add a `depends_on` attribute on each of the key-value- resources.

**Comment by AdamCoulterOz** [comment]
Quoting @arkiaconsulting:

> Thanks @AdamCoulterOz for your contribution.
> 
> One thing that annoys me: I think I will revert back to having the endpoint stay on the key-value resources. Otherwise, the provider configuration is dependent on the App Configuration endpoint, that might be provisioned into the terraform configuration. And because of [this](https://github.com/hashicorp/terraform/issues/2430) a provider configuration cannot reference a resource. Furthermore, that will force us to add a `depends_on` attribute on each of the key-value- resources.

I completely agree, but that it the nature of Terraform. I've been arguing to break the need for discrete separation between plan and apply for years, because it is completely arbitrary.  Case in point - the endpoint, to be in the provider or the resource. We are forced to move it to the resource - just because that's how it is.

@apparentlymart - this problem really needs to be solved
